src/ws-api.ts

```diff
@@ -35,127 +30,14 @@ export class WsApi {
-          );
-
-          const responseBuffer = WsApi.jsonRpcResponseToBuffer(jsonRpcResponse);
-          return socket.send(responseBuffer);
-        }
-
-        dwnRequest = JSON.parse(dwnRequest);
-      } catch (e) {
-        const jsonRpcResponse = createJsonRpcErrorResponse(
-          uuidv4(),
-          JsonRpcErrorCodes.BadRequest,
-          e.message,
-        );
-
-        const responseBuffer = WsApi.jsonRpcResponseToBuffer(jsonRpcResponse);
-        return socket.send(responseBuffer);
-      }
-
-      // Check whether data was provided in the request
-      const { encodedData } = dwnRequest.params;
-      const requestDataStream = encodedData
-        ? DataStream.fromBytes(base64url.baseDecode(encodedData))
-        : undefined;
-
-      const requestContext: RequestContext = {
-        dwn,
-        transport: 'ws',
-        dataStream: requestDataStream,
-      };
-      const { jsonRpcResponse } = await jsonRpcApi.handle(
-        dwnRequest,
-        requestContext,
-      );
-
-      if (jsonRpcResponse.error) {
-        requestCounter.inc({ method: dwnRequest.method, error: 1 });
-      } else {
-        requestCounter.inc({
-          method: dwnRequest.method,
-          status: jsonRpcResponse?.result?.reply?.status?.code || 0,
-        });
-      }
-
-      const responseBuffer = WsApi.jsonRpcResponseToBuffer(jsonRpcResponse);
-      return socket.send(responseBuffer);
-    });
-  }
-
-  /**
-   * This handler returns an interval to ping clients' socket every 30s
-   * if a pong hasn't received from a socket by the next ping, the server will terminate the socket connection.
-   */
-  #setupHeartbeat(): NodeJS.Timer {
-    // Sometimes connections between client <-> server can get borked in such a way that
-    // leaves both unaware of the borkage. ping messages can be used as a means to verify
-    // that the remote endpoint is still responsive. Server will ping each socket every 30s
-    // if a pong hasn't received from a socket by the next ping, the server will terminate
-    // the socket connection
-    return setInterval(() => {
-      this.#wsServer.clients.forEach(function (socket) {
-        if (socket[SOCKET_ISALIVE_SYMBOL] === false) {
-          return socket.terminate();
-        }
-
-        socket[SOCKET_ISALIVE_SYMBOL] = false;
-        socket.ping();
-      });
-    }, HEARTBEAT_INTERVAL);
-  }
-
   /**
    * Handler for starting a WebSocket.
    * Sets listeners for `connection`, `close` events.
    * It clears `heartbeatInterval` when a `close` event is made.
    */
   #setupWebSocket(): void {
-    this.#wsServer.on('connection', this.#handleConnection.bind(this));
-
-    const heartbeatInterval = this.#setupHeartbeat();
-
-    this.#wsServer.on('close', function close() {
-      clearInterval(heartbeatInterval);
-    });
+    this.#wsServer.on('connection', this.#connections.connect.bind(this));
+    this.#wsServer.on('close', this.#connections.close.bind(this));
```

Comment:
Sanity: are you sure the callback function takes any param for `close` event? The original code doesn't take any.

Reply:
Yup! The previous client really didn't have a need for it as there was no cleanup to be done. Now we'd like to close any subscriptions open w/ the DWN to avoid memory leaks, although ping/pong will also handle that.

https://github.com/websockets/ws?tab=readme-ov-file#how-to-detect-and-close-broken-connections

Will include tests for this as well.

Reply:
Ah I misread your comment. You're right I meant to call closeAll here, this isn't for the individual socket, was a bug. Good catch.

Reply:
Fixed